### PR TITLE
Add support for /ignore, /unignore and /ignorelist commands

### PR DESCRIPTION
**Pull request by RockyTV**

### CODE REVIEW

File: src/models/network.js [hunk 1]
```diff
@@ -162,6 +164,8 @@ Network.prototype.export = function() {
 			return _.pick(chan, keys);
 		});
 
+	network.ignoreList = this.ignoreList;
```
Comment (xPaw) [inline src/models/network.js]: This line is unnecessary.

File: src/models/network.js [hunk 2]
```diff
@@ -41,6 +41,7 @@ function Network(attr) {
 			NETWORK: "",
 		},
 		chanCache: [],
+		ignoreList: [],
```
Comment (xPaw) [inline src/models/network.js]: Add `ignoreList` to `filteredFromClient`

File: src/plugins/inputs/ignore.js [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+		target = args[0].trim();
+		hostmask = Helper.parseHostmask(target);
+	}
+
+	switch (cmd) {
+	case "ignore":
+		if (hostmask.nick !== network.nick && !network.ignoreList.some(function(entry) {
+			return (entry.hostmask.nick === hostmask.nick || entry.hostmask.nick === "*") && (entry.hostmask.ident === hostmask.ident || entry.hostmask.ident === "*") && (entry.hostmask.hostname === hostmask.hostname || entry.hostmask.hostname === "*");
+		})) {
+			Array.prototype.push.apply(network.ignoreList, [{
```
Comment (xPaw) [inline src/plugins/inputs/ignore.js]: `network.ignoreList.push` ?

File: test/tests/hostmask.js [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+const Helper = require("../../src/helper");
+
+describe("Hostmask", function() {
+	it("should correctly split hostmask components", function() {
+		expect(Helper.parseHostmask("nick!user@host").nick).to.equal("nick");
+		expect(Helper.parseHostmask("nick!user@host").ident).to.equal("user");
+		expect(Helper.parseHostmask("nick!user@host").hostname).to.equal("host");
+	});
+
+	it("fallback to wildcard if components are empty", function() {
```
Comment (xPaw) [inline test/tests/hostmask.js]: I would like to see more tests for this trying to cover odd cases like empty strings, `!@`, `nick!!!!@thing@@host` and some others.

`compareHostmask` should have some more tests too, where nick is wildcard but hostname is not, etc.

File: src/plugins/inputs/ignore.js [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+				when: Date.now(),
+			});
+
+			client.save();
+		}
+
+		break;
+	case "unignore":
+		network.ignoreList.some(function(entry, idx) {
+			if (Helper.compareHostmask(entry.hostmask, hostmask)) {
```
Comment (xPaw) [inline src/plugins/inputs/ignore.js]: I think unignore needs to only remove exact matches, and not wildcards. Also use `forEach` instead of `some` here.

File: src/helper.js [hunk 6]
```diff
@@ -191,3 +193,43 @@ function mergeConfig(oldConfig, newConfig) {
...
+
+	nick = hostmask || "*";
+
+	const result = {
+		nick: nick,
+		ident: ident,
+		hostname: hostname,
+	};
+
+	result.toString = function() {
```
Comment (xPaw) [inline src/helper.js]: This should be removed, as it more than likely will break after being JSON encoded (e.g. server restart). Just format it where it's needed.

File: src/plugins/inputs/ignore.js [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+	if (cmd !== "ignorelist") {
+		// Trim to remove any spaces from the hostmask
+		target = args[0].trim();
+		hostmask = Helper.parseHostmask(target);
+	}
+
+	switch (cmd) {
+	case "ignore":
+		if (hostmask.nick !== network.nick && !network.ignoreList.some(function(entry) {
```
Comment (xPaw) [inline src/plugins/inputs/ignore.js]: Send an error message when you try to ignore yourself instead of silently doing nothing.

It should also send an error when you try to ignore someone that is already ignored.

File: src/plugins/inputs/ignore.js (hunk 5 above)
Comment (xPaw) [reply src/plugins/inputs/ignore.js]: Also add messages depending on if unignore worked or not.

File: src/plugins/inputs/ignore.js [hunk 8]
```diff
@@ -0,0 +1,89 @@
...
+			return Helper.compareHostmask(entry.hostmask, hostmask);
+		})) {
+			network.ignoreList.push({
+				hostmask: hostmask,
```
Comment (xPaw) [inline src/plugins/inputs/ignore.js]: Instead of creating a sub object, I would just add `hostmask.when` so it's flat.

File: test/tests/hostmask.js [hunk 9]
```diff
@@ -0,0 +1,25 @@
+"use strict";
+
+const expect = require("chai").expect;
+const Helper = require("../../src/helper");
+
+describe("Hostmask", function() {
```
Comment (astorije) [inline test/tests/hostmask.js]: Probably worth colocating the tests in https://github.com/thelounge/thelounge/blob/master/test/src/helperTest.js, and using `describe("#parseHostmask", function() {` (actually should be `".parseHostmask"`).
Thoughts?

Comment (xPaw) [reply test/tests/hostmask.js]: I really don't mind these function tests being in their own file (a couple of other functions are already separate)

Comment (astorije) [reply test/tests/hostmask.js]: Fair enough. At least we should split them per function though.

**Comment by McInkay** [comment]
I'd like to block this until #1878 is merged. Hopefully shouldn't be too long as all current comments have been addressed and it's just waiting for a last review from @astorije

File: src/plugins/irc-events/message.js [hunk 10]
```diff
@@ -41,13 +42,23 @@ module.exports = function(irc, network) {
 		let from;
 		let highlight = false;
 		let showInActive = false;
+		let shouldIgnore = false;
 		const self = data.nick === irc.user.nick;
 
+		// Check if the sender is in our ignore list
+		shouldIgnore = network.ignoreList.some(function(entry) {
```
Comment (xPaw) [inline src/plugins/irc-events/message.js]: You can do `const shouldIgnore = ...` and move it just before `if` where this variable is relevant.

File: src/plugins/inputs/ignore.js [hunk 11]
```diff
@@ -0,0 +1,117 @@
...
+
+	if (cmd !== "ignorelist") {
+		// Trim to remove any spaces from the hostmask
+		target = args[0].trim();
+		hostmask = Helper.parseHostmask(target);
+	}
+
+	switch (cmd) {
+	case "ignore": {
+		if (hostmask.nick === network.nick) {
```
Comment (xPaw) [inline src/plugins/inputs/ignore.js]: This should be case insensitive check. You can just add `toUpperCase` on both strings.

File: src/helper.js [hunk 12]
```diff
@@ -191,3 +193,40 @@ function mergeConfig(oldConfig, newConfig) {
...
+	parts = hostmask.split("!");
+
+	if (parts.length >= 2) {
+		ident = parts[1] || "*";
+		hostmask = parts[0];
+	}
+
+	nick = hostmask || "*";
+
+	const result = {
```
Comment (xPaw) [inline src/helper.js]: As you asked on IRC, hostmasks indeed also need to be compared case insensitively. I think it's pretty safe to lowercase it in parse function (instead of in compare function all the time).

Don't forget to add tests for lowercasing :)

File: src/plugins/irc-events/message.js [hunk 13]
```diff
@@ -41,13 +42,23 @@ module.exports = function(irc, network) {
...
+		});
+
 		// Server messages go to server window, no questions asked
 		if (data.from_server) {
 			chan = network.channels[0];
 			from = chan.getUser(data.nick);
 		} else {
+			if (shouldIgnore) {
+				return;
+			}
```
Comment (astorije) [inline src/plugins/irc-events/message.js]: So messages from ignored people are trashed on reception, correct?

I explained that differently [here](https://github.com/thelounge/thelounge/issues/464#issuecomment-373262236), but I think we should not completely delete based ignores on reception. When ignoring someone, I think all their previous messages should be removed too, and when unignoring someone, all messages should re-appear.
Thoughts?

Comment (McInkay) [reply src/plugins/irc-events/message.js]: I agree. Messages shouldn't disappear into the aether. They should be set as ignored somehow on the message and things that ignore them (highlighting, displaying) should use that feature.

Comment (xPaw) [reply src/plugins/irc-events/message.js]: That's gonna be much more painful to implement what you are suggesting.

Comment (RockyTV) [reply src/plugins/irc-events/message.js]: Okay, not wanting to use a fallacy, but every other IRC client just ignores the messages after the mask has been added to the list. And when you remove the mask from the list, previous messages are not shown.

I think there's no easy way to do what you guys are asking for, pretty sure it'd need to make the client aware of these messages and just apply a CSS filter to them or something like that.

Comment (dgw) [reply src/plugins/irc-events/message.js]: However painful to implement, it's the better UX. I wouldn't say the first iteration of ignore support needs to work that way, since several other IRC clients I've used just irrevocably discard ignored messages. But it's a good goal to have `/ignore` and `/unignore` work retroactively.

Comment (McInkay) [reply src/plugins/irc-events/message.js]: Yeah, having chatted with @xPaw on IRC, I agree that it would be fine to have this for the first version, but we want to work towawrds it after this PR.

@astorije ?

File: src/helper.js [hunk 14]
```diff
@@ -191,3 +193,44 @@ function mergeConfig(oldConfig, newConfig) {
...
+
+	ident = ident.toLowerCase();
+
+	nick = hostmask.toLowerCase() || "*";
+
+	const result = {
+		nick: nick,
+		ident: ident,
+		hostname: hostname,
+		when: new Date(1970, 0, 1),
```
Comment (xPaw) [inline src/helper.js]: You don't need this line.

File: src/plugins/irc-events/message.js (hunk 13 above)
Comment (astorije) [reply src/plugins/irc-events/message.js]: That's fair. Your PR has a lot of useful work around managing ignores, we can improve the actual ignoring later (assuming we will though).

I don't think there needs to be CSS involved though: when a message is received, it is instantiated and added to the channel, but not sent to the client. When a client loads channels, messages are compared to the ignored list and those that should be ignored are not sent to the client.

Of course, it's more complex that just dropping ignored messages, as there will be more edge cases (dealing with messages that usually open a new query window, making this work without reloading the page, etc.) but that can be improved over time :)

**Comment by Adrian-Turjak** [comment]
Can we possibly add slightly more granularity for the ignore functionality when it comes to joins parts, quits? So for channels, and users, having the option to ignore specific interaction types you see, would be good.

E.g. I want to see joins and such for most stuff, but some bot users I don't since they leave/join all the time.

I tried doing this with css (which would work for me), but joins parts and quits do not have data-from for the user they are related to... so I can't.

**Comment by McInkay** [comment]
@Adrian-Turjak Yes, but that'll come later. This is a good start.

**Comment by McInkay** [comment]
This will need to be refactored now that #1878 is merged, and use the new structure for adding context menu items. If you have any questions about how this works, or anything is unclear about what I've changed then give me a shout @RockyTV, either here or on irc.

**Comment by Adrian-Turjak** [comment]
@xPaw 
> You must be using an outdated version, because actions do have data-from set.

We likely are! Will try when we upgrade. We're currently waiting for the next release with the log reloading.

**Comment by astorije** [comment]
Hey @Adrian-Turjak, just curious, when you say "we", do you mean your company? Are you guys using The Lounge for work?

**Comment by Adrian-Turjak** [comment]
@astorije Running it for our team at work on a spare workstation machine with coreos under someone's desk. It's just being used for maybe 6 or so people.

**Comment by astorije** [comment]
@Adrian-Turjak, that's pretty cool! Any chance I could contact you in private (email) to ask you a few follow-up questions on your experience? If you don't want to expose your email on GH, feel free to drop me an email at jeremie(at)astori.fr.

@McInkay, I believe @RockyTV is done updating this PR, so it's open for review again!

**Comment by RockyTV** [comment]
Fixed the merge conflict. Is there anything else I should change or do? If not, then it's ready to be reviewed.

File: src/client.js [hunk 15]
```diff
@@ -46,6 +46,7 @@ const inputs = [
 	"away",
 	"connect",
 	"disconnect",
+	"ignore",
```
[the remark this answers is not in the document]
Comment (RockyTV) [reply src/client.js]: They're in `ignore.js`, which is a single input file.

[the remark this answers is not in the document]
Comment (McInkay) [reply src/client.js]: Oh duh, I was thinking this was the list of commands for autocompletion locally, but that's not what this is...sigh. Ignore me, thanks.

**Comment by astorije** [comment]
@xPaw, mind reviewing that? I can also give it a light review after you're okay with it if you want.

File: src/plugins/inputs/ignore.js [hunk 16]
```diff
@@ -0,0 +1,120 @@
...
+			}));
+		} else if (!network.ignoreList.some(function(entry) {
+			return Helper.compareHostmask(entry, hostmask);
+		})) {
+			hostmask.when = Date.now();
+			network.ignoreList.push(hostmask);
+
+			client.save();
+			chan.pushMessage(client, new Msg({
+				type: Msg.Type.ERROR,
```
Comment (astorije) [inline src/plugins/inputs/ignore.js]: This shouldn't be an error. Is there any other type we could use?

File: src/plugins/inputs/ignore.js [hunk 17]
```diff
@@ -0,0 +1,120 @@
...
+		});
+
+		// Check if the entry exists before removing it, otherwise
+		// let the user know.
+		if (idx !== -1) {
+			network.ignoreList.splice(idx, 1);
+			client.save();
+
+			chan.pushMessage(client, new Msg({
+				type: Msg.Type.ERROR,
```
Comment (astorije) [inline src/plugins/inputs/ignore.js]: Same here.

**Comment by astorije** [comment]
Testing this when rebased on `master`, clicking on "List ignored users" or sending `/ignorelist` doesn't do anything (even after I added someone to the ignore list). Is there something I'm missing?

**Comment by RockyTV** [comment]
@astorije I have double-checked the code and I don't think there's an issue with it. I don't know what happened since last month but lists are now broken, i.e., ban list, ignore list and the channel list aren't being displayed by TL.

**Comment by RockyTV** [comment]
Should be good to go.

**Comment by astorije** [comment]
Alright, testing: `The Lounge source (e81acc4d / v3.0.0-pre.6) (Node.js 8.9.4 on darwin x64)`

`/ignorelist` now works, but right clicking on network and clicking on "List ignored users" does nothing. @RockyTV or @xPaw, could you re-test on your end maybe?

**Comment by RockyTV** [comment]
@astorije Sorry, fixed it now. Should be fine.

File: src/plugins/inputs/ignore.js [hunk 18]
```diff
@@ -0,0 +1,120 @@
...
+				text: "Successfully removed user/hostmask from ignorelist",
```
Comment (xPaw) [inline src/plugins/inputs/ignore.js]: Can we print which hostmask was removed?

File: src/plugins/irc-events/message.js (hunk 13 above)
Comment (zchrykng) [reply src/plugins/irc-events/message.js]: It would be nice if it worked similarly to weechat's filters. You can toggle those on and off and all matching messages are hidden/shown depending on the state of the filter.
